### ISSUE TITLE
Allow Firewalls to be added by a bundle

### ISSUE THREAD

**Issue by samuel-sol** [symfony/symfony#51487]
### Description

Hey symfony group, I'm developing an internal set of tools based on symfony. These tools have a shared bundle that adds a bunch of api calls that will use a separate firewall to authenticate (cert based). But when I tried to prepend  the config on the bundle I get the following message 

`You are not allowed to define new elements for path "security.firewalls". Please define all elements for this path in one config file.`

I saw a couple of years old thread about it, but they were abandoned, is there a reason to not allow firewall to be used on other places?

**Comment by stof** [comment]
The first file defining firewall must define all the firewalls, because the order of firewalls is important (the first matching firewall is the one being used) and merging new values from later files would not allow to control where the new firewall is inserted in the list.

It is possible for later configurations to contribute to the configuration of firewalls however, as long as they configure settings for firewalls that are already known (and so already sorted).

**Comment by samuel-sol** [comment]
> The first file defining firewall must define all the firewalls, because the order of firewalls is important (the first matching firewall is the one being used) and merging new values from later files would not allow to control where the new firewall is inserted in the list.

Doesn't the prepend extension  method guarantee that you add on the start of the configuration?  That way you guarantee that whatever you are adding is first on the config.

**Comment by stof** [comment]
@samuel-sol I'm talking about the order of firewalls in the final SecurityBundle configuration, not of the order of configurations in the input of the extension.
An important feature of processing the configuration of the bundle is merging multiple sources of config (coming either from different files configuring the same bundle or even from different sections in the same file thanks to the `when@prod` syntax) in order to produce a single configuration (another important feature of this processing is of course to validate the structure of the provided configuration).

When using `prepend`, the configuration you add indeed becomes the first one in the input (temporarily as some other bundle could prepend another config after that, taking over the first position...). And that's precisely the issue in your case IMO. The first configuration containing a configuration for security firewalls is expected to provide the sorted list of firewall names (later files can override some settings in the known keys for firewalls, as we don't need to do merging in the list itself in such case but only deeper in the config of that firewall, which does not have issues with sorting anymore). I suspect that your prepended config contains only 1 firewall.

**Comment by samuel-sol** [comment]
>  I suspect that your prepended config contains only 1 firewall.

Basically. Here is my scenario. I have two applications, will be three in a while. Each application has one firewall tied to a user provider and a couple of authentication methods.  This firewall is defined to ignore anything that starts with `/api` . 

```
    main:
      pattern: ^/(?!api)(?:/.+)?$
      lazy: true
      entry_point: form_login
      provider: app_user_provider
```


Both applications will have access to a REST API that is being developed as a bundle.  

```
    first_api:
      pattern: ^/api/first/.+$
      security: true
      lazy: true
```

I think that if all your firewalls have a  pattern, we could allow to be added in multiple files since they don't "clash".

**Comment by stof** [comment]
even if all firewalls have a pattern, it does not mean that all firewalls have exclusive patterns for which the matching order is irrelevant.
The most common configuration is to have a catch-all pattern for the main firewall defined at the end of the list, while some more specific patterns would match before that. That's why the order of firewalls is very important.

Forcing the order of firewalls to be defined explicitly avoids bugs where the merged order would not be the one that you would expect, as that could become a security issue in your app if the wrong firewall is used for a URL.
And trying to check whether patterns are exclusive to each other is very hard (it would require analyzing 2 regexps and determining whether there can be a string matching both, without any restriction on the regexp features being used). Thus, by the time we have to merge 2 configurations with different firewall keys, we have no guarantee that the pattern is already defined at its final value (as more configurations can be merged later).

**Comment by samuel-sol** [comment]
> And trying to check whether patterns are exclusive to each other is very hard 

Oh for sure, I think it would be on the developer to enforce it, no way to do it programmatically. One possibility I thought would be adding a `priority`  option on the firewall configuration. If present the firewalls could be sorted by it. If not present it works as it is now. And make it mandatory IF there are firewalls configured in multiple files. This would be backwards compatible and I think it would allow this functionality without many issues.

So basically, you would have a couple of options: 

- Same file and no priority: `default`, works like it is right now
- Same file, all firewalls have a `priority`: sort by priority instead of file order 
- Multiple files, all firewalls have a `priority`: sort by priority

**Comment by stof** [comment]
there is no (maintainable) way we can do what you suggest, as the list merging has no knowledge of the deeper keys (not even mentioning that any deeper key is not final yet by the time the list merging happens).

We could go with a `priority` field as done in monolog-bundle for the handlers, with a default priority of 0, which would keep items defined _for their first time_ in the same file in their relative order (thanks to stable sorting), but we would not be able to warn you if some merging happens with no priority being defined. And you might still not realize that the actual merged list is in the wrong order, introducing a security issue, even if the priority would provide a tool to fix it. Having log handlers in a wrong order is much less risky than having security firewalls in the wrong order.

**Comment by samuel-sol** [comment]
> there is no (maintainable) way we can do what you suggest, as the list merging has no knowledge of the deeper keys (not even mentioning that any deeper key is not final yet by the time the list merging happens).

Ah, I didn't know about this, thanks for explaining.

>  And you might still not realize that the actual merged list is in the wrong order, introducing a security issue, even if the priority would provide a tool to fix it.

Why exactly is that a security issue? In case a random package removes authentication? 

One option I see would be adding a `allow_external_firewalls` key to security.firewalls. Defaulting it to false, and them you have to explicity alllow external merges and that would allow sorting by priority. Forcing you to explicitly allow the merge. Could that be checked when the config is being joined?

**Comment by stof** [comment]
We cannot rely on the semantic configuration to control how the semantic configuration is processed. That would be a chicken-&-egg issue (and an external bundle could be the one setting that key...)

**Comment by wouterj** [comment]
Besides the technical limitations, I also don't feel comfortable with bundles adding firewalls. In general, bundles should not interfere in the infrastructure side of an application, but leave this up to the application using the bundle. By design, this makes the bundles more sharable across applications (e.g. what if application X wants the shared API calls under a different path?).

An example of a severe security issue would be a bundle that defines a firewall for `^/` with `security: false`. Furthermore, firewalls also need to define the user provider. So if we give bundles access to define firewalls, we must also allow them to add user providers. This would increase the attack surface for supply chain attacks even more.
